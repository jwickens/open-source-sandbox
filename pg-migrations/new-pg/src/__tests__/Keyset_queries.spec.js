/**
 * @flow
 */
const { Db } = require('../')
const path = require('path')

const OPTIONS = {
  schemaBase: 'keysetquerytests',
  schemaVersion: 0.3,
  overrideEnv: true,
  sqlDirectory: path.join(__dirname, 'sql')
}

const db = new Db(OPTIONS)
let table
beforeAll(async () => {
  await db.sync()
  await db.migrateTo(1)
  table = db.getTable('test')
})

afterAll(async () => {
  await db.close()
})

describe('Keyset', () => {
  test('can get a paginated list (basic)', async () => {
    const keyset = table.getKeyset()
      .select(['id', 'tags', 'message'])
      .start({field: 'id', first: 1})
    const result = await keyset.query()
    expect(result).toMatchSnapshot()
    const plan = await db.query(`explain ${keyset.toSql()}`)
    expect(keyset.toSql()).toMatchSnapshot()
    expect(plan.rows.map(r => r['QUERY PLAN']).join('\n')).toMatchSnapshot()
  })

  test.skip('can get a paginated list (search)', async () => {
    const keyset = table.getKeyset()
      .select(['id', 'tags', 'message'])
      .start({field: 'message', searchQuery: 'hell', first: 1})

    const result = await keyset.query()
    expect(result.edges[0].node.message).toMatch('hell')
    expect(result).toMatchSnapshot()
    const plan = await db.query(`explain ${keyset.toSql()}`)
    expect(keyset.toSql()).toMatchSnapshot()
    expect(plan.rows.map(r => r['QUERY PLAN']).join('\n')).toMatchSnapshot()
  })

  test('can start from the end', async () => {
    const keyset = table.getKeyset()
      .select(['id', 'tags', 'message'])
      .start({field: 'id', last: 10})
    const result = await keyset.query()
    expect(result.pageInfo).toEqual({hasNextPage: true, hasPrevPage: false, totalCount: 463})
    expect(result).toMatchSnapshot()
    const plan = await db.query(`explain ${keyset.toSql()}`)
    expect(keyset.toSql()).toMatchSnapshot()
    expect(plan.rows.map(r => r['QUERY PLAN']).join('\n')).toMatchSnapshot()
  })

  test('can use default last option', async () => {
    const keyset = table.getKeyset()
      .select(['id', 'tags', 'message'])
      .start({field: 'id', defaultLast: true})
    const result = await keyset.query()
    expect(result.pageInfo).toEqual({hasNextPage: true, hasPrevPage: false, totalCount: 463})
    expect(result).toMatchSnapshot()
    const plan = await db.query(`explain ${keyset.toSql()}`)
    expect(keyset.toSql()).toMatchSnapshot()
    expect(plan.rows.map(r => r['QUERY PLAN']).join('\n')).toMatchSnapshot()
  })
})
