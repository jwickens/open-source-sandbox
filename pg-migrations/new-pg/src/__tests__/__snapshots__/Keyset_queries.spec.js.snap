// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyset can get a paginated list (basic) 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "eyJzIjpbeyJ2IjoxLCJmIjoiaWQiLCJkIjpmYWxzZX1dLCJmIjpbXSwidyI6W10sImkiOltdLCJuIjpbXX0=",
      "node": anonymous {
        "id": 1,
        "message": "these messages have a default",
        "tags": Array [
          "a",
        ],
      },
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPrevPage": false,
    "totalCount": 463,
  },
}
`;

exports[`Keyset can get a paginated list (basic) 2`] = `
"
      SELECT id, tags, message
      FROM test
      WHERE id is not null
      
      ORDER BY id ASC
      LIMIT 1
    "
`;

exports[`Keyset can get a paginated list (basic) 3`] = `
"Limit  (cost=0.28..0.36 rows=1 width=68)
  ->  Index Scan using test_pkey on test  (cost=0.28..86.02 rows=1014 width=68)
        Index Cond: (id IS NOT NULL)"
`;

exports[`Keyset can get a paginated list (search) 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "eyJzIjpbeyJ2IjowLjIxMDUyNiwiZiI6Im1lc3NhZ2UiLCJkIjp0cnVlLCJxIjoiaGVsbCJ9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 427,
        "message": "hello there c tags",
        "sml": 0.210526,
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPrevPage": false,
  },
}
`;

exports[`Keyset can get a paginated list (search) 2`] = `
"
      SELECT id, tags, message, similarity(message, 'hell') as sml
      FROM test
      WHERE message is not null AND message like '%hell%'
      ORDER BY similarity(message, 'hell') DESC
      LIMIT 1
    "
`;

exports[`Keyset can get a paginated list (search) 3`] = `
"Limit  (cost=23.80..23.80 rows=1 width=72)
  ->  Sort  (cost=23.80..23.82 rows=8 width=72)
        Sort Key: (similarity(message, 'hell'::text)) DESC
        ->  Seq Scan on test  (cost=0.00..23.76 rows=8 width=72)
              Filter: ((message IS NOT NULL) AND (message ~~ '%hell%'::text))"
`;

exports[`Keyset can start from the end 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjMsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 463,
        "message": "hello there c tags",
        "tags": Array [
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjIsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 462,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjEsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 461,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjAsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 460,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTksImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 459,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTgsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 458,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTcsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 457,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTYsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 456,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTUsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 455,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTQsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 454,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPrevPage": false,
    "totalCount": 463,
  },
}
`;

exports[`Keyset can start from the end 2`] = `
"
      SELECT id, tags, message
      FROM test
      WHERE id is not null
      
      ORDER BY id DESC
      LIMIT 10
    "
`;

exports[`Keyset can start from the end 3`] = `
"Limit  (cost=0.28..1.12 rows=10 width=68)
  ->  Index Scan Backward using test_pkey on test  (cost=0.28..86.02 rows=1014 width=68)
        Index Cond: (id IS NOT NULL)"
`;

exports[`Keyset can use default last option 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjMsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 463,
        "message": "hello there c tags",
        "tags": Array [
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjIsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 462,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjEsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 461,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NjAsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 460,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTksImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 459,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTgsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 458,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTcsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 457,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTYsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 456,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTUsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 455,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTQsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 454,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTMsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 453,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTIsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 452,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTEsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 451,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NTAsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 450,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NDksImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 449,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NDgsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 448,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NDcsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 447,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NDYsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 446,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NDUsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 445,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
    Object {
      "cursor": "eyJzIjpbeyJ2Ijo0NDQsImYiOiJpZCIsImQiOnRydWV9XSwiZiI6W10sInciOltdLCJpIjpbXSwibiI6W119",
      "node": anonymous {
        "id": 444,
        "message": "hello there c tags",
        "tags": Array [
          "a",
          "c",
        ],
      },
    },
  ],
  "pageInfo": Object {
    "hasNextPage": true,
    "hasPrevPage": false,
    "totalCount": 463,
  },
}
`;

exports[`Keyset can use default last option 2`] = `
"
      SELECT id, tags, message
      FROM test
      WHERE id is not null
      
      ORDER BY id DESC
      LIMIT 20
    "
`;

exports[`Keyset can use default last option 3`] = `
"Limit  (cost=0.28..1.97 rows=20 width=68)
  ->  Index Scan Backward using test_pkey on test  (cost=0.28..86.02 rows=1014 width=68)
        Index Cond: (id IS NOT NULL)"
`;
